File: 2026/09-September/16-Wednesday/solution.py
class Solution:
    def numberOfSets(self, n: int, k: int) -> int:
        MOD = 10**9 + 7

        # dp[i][j] = number of ways using first i points
        # to create j non-overlapping segments.
        dp = [[0] * (k + 1) for _ in range(n)]

        for i in range(n):
            dp[i][0] = 1

        for i in range(1, n):
            for j in range(1, k + 1):
                # Don't use point i as the end of a new segment
                dp[i][j] = dp[i - 1][j]

                # Choose the starting point of the new segment.
                # Summing over all possible starts:
                for start in range(i):
                    dp[i][j] += dp[start][j - 1]
                    dp[i][j] %= MOD

        return dp[n - 1][k]

File: 2026/09-September/16-Wednesday/solution.java
class Solution {
    public int numberOfSets(int n, int k) {
        final int MOD = 1_000_000_007;

        long[][] dp = new long[n][k + 1];

        // 0 segments can always be formed in exactly 1 way
        for (int i = 0; i < n; i++) {
            dp[i][0] = 1;
        }

        for (int i = 1; i < n; i++) {
            for (int j = 1; j <= k; j++) {

                // Don't use point i as the endpoint of a segment
                dp[i][j] = dp[i - 1][j];

                // Choose the starting point of the new segment
                for (int start = 0; start < i; start++) {
                    dp[i][j] = (dp[i][j] + dp[start][j - 1]) % MOD;
                }
            }
        }

        return (int) dp[n - 1][k];
    }
}
